### PR TITLE
ARROW-11790: [Rust][DataFusion] Change builder signatures to take impl Interator<Item=Expr> rather than &[Expr] 

### PR DESCRIPTION
**Pull request by alamb**
# NOTE:
Since is a fairly major backwards incompatible change (many callsites need to be updated, though mostly mechanically); I gathered some feedback on this approach in https://github.com/apache/arrow/pull/9692 and this is the PR I propose for merge. 

I'll leave this open for several days and also send a note to the mailing lists for additional comment

It is part of my overall plan to make the DataFusion optimizer more idiomatic and do much less copying [ARROW-11689](https://issues.apache.org/jira/browse/ARROW-11689)

# Rationale:
All callsites currently need an owned `Vec` (or equivalent) so they can pass in `&[Expr]` and then Datafusion copies all the `Expr`s. Many times the original `Vec<Expr>` is discarded immediately after use (I'll point out where this happens in a few places below). Thus I it would better (more idiomatic and often less copy/faster) to take something that could produce an iterator over Expr

# Changes
1. Change `Dataframe` so it takes `Vec<Expr>` rather than `&[Expr]`
2. Change `LogicalPlanBuilder` so it takes `impl Iterator<Item=Expr>` rather than `&[Expr]`

I couldn't figure out how to allow the `Dataframe` API (which is a Trait) to take an `impl Iterator<Item=Expr>`

### CODE REVIEW

File: rust/benchmarks/src/bin/tpch.rs [hunk 1]
```diff
@@ -1636,7 +1636,7 @@ mod tests {
                 .file_extension(".out");
             let df = ctx.read_csv(&format!("{}/answers/q{}.out", path, n), options)?;
             let df = df.select(
-                &get_answer_schema(n)
```
Comment (alamb) [inline rust/benchmarks/src/bin/tpch.rs:1639]: this is a pretty good example -- the `answer_schema` was created and immediately dropped after calling `df.select()`. Now the exprs don't need to be copied

File: rust/datafusion/src/logical_plan/builder.rs [hunk 2]
```diff
@@ -39,6 +39,43 @@ use crate::logical_plan::{DFField, DFSchema, DFSchemaRef, Partitioning};
...
+/// #    Schema::new(vec![
+/// #           Field::new("id", DataType::Int32, false),
+/// #           Field::new("first_name", DataType::Utf8, false),
+/// #           Field::new("last_name", DataType::Utf8, false),
+/// #           Field::new("state", DataType::Utf8, false),
+/// #           Field::new("salary", DataType::Int32, false),
+/// #       ])
+/// #   }
+/// #
+/// // Create a plan similar to
```
Comment (alamb) [inline rust/datafusion/src/logical_plan/builder.rs:61]: As I initially found the `impl IntoIterator<Item ....` syntax hard to grok, I figured I would add an example to `LogicalPlanBuilder` showing that it lets you use `vec![]` and the trait business can be somewhat ignored

File: rust/datafusion/src/logical_plan/builder.rs [hunk 3]
```diff
@@ -220,20 +260,28 @@ impl LogicalPlanBuilder {
         }))
     }
 
-    /// Apply an aggregate
-    pub fn aggregate(&self, group_expr: &[Expr], aggr_expr: &[Expr]) -> Result<Self> {
-        let mut all_expr = group_expr.to_vec();
```
Comment (alamb) [inline rust/datafusion/src/logical_plan/builder.rs:225]: it was even worse here -- *two* copies of each input expr were made -- one copy to pass to `validate_unique_names` and once to put on the `LogicalPlanAggregate`.

Admittedly, removing the second copy doesn't require changing the signature of this function

**Comment by returnString** [comment]
> I couldn't figure out how to allow the Dataframe API (which is a Trait) to take an impl Iterator<Item=Expr>

Yeah, it's a pain but `impl trait` (in both return and argument position) isn't supported in traits right now. Afaik there were some historical discussions about opening up the system to support it but I can't find anything recent.

Related: is there a reason that `DataFrame` is a trait with a single impl? I'm wondering if it could be a concrete struct with the actual execution logic handling more exotic setups (custom `TableProvider` impls etc), which would free it from all the trait-imposed restrictions. Are we expecting to support multiple DF implementations?

**Comment by alamb** [comment]
> Related: is there a reason that DataFrame is a trait with a single impl? I'm wondering if it could be a concrete struct with the actual execution logic handling more exotic setups (custom TableProvider impls etc), which would free it from all the trait-imposed restrictions. Are we expecting to support multiple DF implementations?

@jorgecarleitao  might know the answer to that as he introduced the `DataFrame` trait initially. @ritchie46  has a dataframe library polars which maybe someday could be integrated: https://www.ritchievink.com/blog/2021/02/28/i-wrote-one-of-the-fastest-dataframe-libraries/. It is interesting to me that many of the polars features and implementation seems very similar to the  features in `DataFusion`

**Review by returnString — APPROVED** [review]

**Comment by jorgecarleitao** [comment]
@alamb , I do not know the reason. I did not use it on `datafusion-python`, mostly because the trait is not really used anywhere: we have no generics or `dyn DataFrame`. I am fine dropping it until we see a need for it, but @andygrove likely had some plans here :)

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Thanks a lot, @alamb .

**Comment by andygrove** [comment]
> @alamb , I do not know the reason. I did not use it on `datafusion-python`, mostly because the trait is not really used anywhere: we have no generics or `dyn DataFrame`. I am fine dropping it until we see a need for it, but @andygrove likely had some plans here :)

I can explain :-)

My motivation was extensibility. I wanted users to be able to use the DataFusion DataFrame for local (in-process) execution but then be able to run distributed by having a mechanism to tell the context which DataFrame impl to use (configuration change rather than code change).

I have a much better idea now though of the requirements so plan on writing up a design once Ballista is in this repo.

**Comment by andygrove** [comment]
And just to be clear, I am fine with removing the trait.

**Comment by alamb** [comment]
I am going to rebase this PR and merge it once CI passes

**Comment by alamb** [comment]
> And just to be clear, I am fine with removing the trait.

Thanks @andygrove  -- I have no particular need to remove the `DataFrame` trait, but if it gets in the way we can definitely revisit 👍

**Comment by andygrove** [comment]
I wrote up some very brief notes in https://issues.apache.org/jira/browse/ARROW-12064

**Comment by jorgecarleitao** [comment]
Got. Then fwiw, I think that we can build that with this trait, we just need a bit more work / investigation:

I am not 100% sure, but I think that the path for libraries that want to expose a different `DataFrame` (datafusion-python needs this) and use datafusion's `ExecutionContext` is to make:

`struct ExecutionContext<D: DataFrame>`

For the case of Ballista, I think we need the opposite: Ballista has an `ExecutionContext`, and `DataFrame` should just have a way of absorving that implementation, e.g. via `DataFrame<C: ExecutionContext>` (we need a trait for `ExecutionContext` for this).

(This is a hand-waving idea)

**Review by houqp — APPROVED** [review]
very nice refactor

**Pull request closed without merging**